Testing/SystemTests/tests/framework/DocumentationTest.py

```diff
@@ -0,0 +1,161 @@
+    "DGS Reduction.rst",  # reduction application
+    "Direct Data Analysis.rst",  # opens Inelastic Data Analysis
+    "Indirect Data Analysis.rst",  # opens Inelastic Data Analysis
+    "HFIR 4Circle Reduction.rst",
+]
+
+
+def _file_exists_under_dir(dir_path: Path, file_name: str) -> bool:
+    """
+    Search for file name in the given directory and all its subdirectories
+    :param dir_path: directory to search under
+    :param file_name: name of the file
+    :return: bool, true if file found
+    """
+    found_files = [path for path in dir_path.rglob(file_name) if path.is_file]
+    return len(found_files) > 0
+
+
+def _get_interface_names() -> Dict[str, List[str]]:
+    # need to initialise this so cpp interfaces are registered
+    InterfaceManager()
+    interfaces = gather_interface_names()
+    hidden_interfaces = ConfigService["interfaces.categories.hidden"].split(";")
+    interface_names = {key: names for key, names in interfaces.items() if key not in hidden_interfaces}
+    return interface_names
+
+
+class AlgorithmDocumentationTest(MantidSystemTest):
+    """
+    This test checks that all registered algorithms have a documentation page.
+    """
+
+    def runTest(self):
+        all_algs = AlgorithmFactory.getRegisteredAlgorithms(False)
+        self.assertGreaterThan(len(all_algs), 0, msg="No registered algorithms found")
+        missing = []
+
+        for name, versions in all_algs.items():
+            for version in versions:
+                file_name = f"{name}-v{version}.rst"
+                if not _file_exists_under_dir(ALG_DOCS_PATH, file_name):
+                    if file_name not in ALGORITHMS_WITH_NO_DOCS:
+                        missing.append(file_name)
+                elif file_name in ALGORITHMS_WITH_NO_DOCS:
+                    raise FileExistsError(f"{file_name} exists but is still in the exceptions list. Please update the list.")
+
+        self.assertEqual(len(missing), 0, msg="Missing the following algorithm docs:\n" + "\n".join(missing))
+
+
+class FittingDocumentationTest(MantidSystemTest):
+    """
+    This test checks that all registered fit functions have a documentation page.
+    """
+
+    def runTest(self):
+        all_fit_funcs = FunctionFactory.getFunctionNames()
+        self.assertGreaterThan(len(all_fit_funcs), 0, msg="No available fit functions found")
+        missing = []
+
+        for name in all_fit_funcs:
+            file_name = f"{name}.rst"
+            if not _file_exists_under_dir(FIT_DOCS_PATH, file_name):
+                if file_name not in FIT_FUNCTIONS_WITH_NO_DOCS:
+                    missing.append(file_name)
+            elif file_name in FIT_FUNCTIONS_WITH_NO_DOCS:
+                raise FileExistsError(f"{file_name} exists but is still in the exceptions list. Please update the list.")
+
+        self.assertEqual(len(missing), 0, msg="Missing the following fit function docs:\n" + "\n".join(missing))
+
+
+class InterfaceDocumentationTest(MantidSystemTest):
```

Comment:
These individual system test classes might benefit from going into their own files. There is already a lot going on in these system tests, so moving them into their separate files might help with readability

Reply:
It might be possible to do something like the following (a similar thing is done for the `AlgorithmValidateInputs` system tests):

```python
# Define a base test class in a separate file
class DocumentationTestBase(MantidSystemTest, metaclass=ABCMeta):

    def __init__(self):
        super(DocumentationTestBase, self).__init__()
        self._no_docs = []
        self._pages = []
        self._test_type = ""
        self._doc_path = ""
        self._setup_test()

    @abstractmethod
    def _setup_test(self):
        raise NotImplementedError("This method needs implementing in a subclass of this class.")

    def runTest(self):
        all_algs = AlgorithmFactory.getRegisteredAlgorithms(False)
        self.assertGreaterThan(len(self._pages), 0, msg=f"No registered {self._test_type} found")
        missing = []

        for name, versions in self._pages.items():
            for version in versions:
                file_name = f"{name}-v{version}.rst"
                if not _file_exists_under_dir(self._doc_path, file_name):
                    if file_name not in self._no_docs:
                        missing.append(file_name)
                elif file_name in self._no_docs:
                    raise FileExistsError(f"{file_name} exists but is still in the exceptions list. Please update the list.")

        self.assertEqual(len(missing), 0, msg=f"Missing the following {self._test_type} docs:\n" + "\n".join(missing))


# import this base class into each of your documentation test files
class AlgorithmDocumentationTest(DocumentationTestBase):


    def _setup_test(self) -> None:
        self._no_docs = ALGORITHMS_WITH_NO_DOCS 
        self._pages = AlgorithmFactory.getRegisteredAlgorithms(False)
        self._test_type = "Algorithm"
        self._doc_path = <insert path to docs here>

```

A bit like this https://github.com/mantidproject/mantid/blob/81c91daaa5b951f60631945bc2b9e8d7278c84c8/Testing/SystemTests/tests/framework/AlgorithmValidateInputsTestBase.py#L15

https://github.com/mantidproject/mantid/blob/81c91daaa5b951f60631945bc2b9e8d7278c84c8/Testing/SystemTests/tests/framework/AlgorithmMatrixWorkspaceValidateInputsTest.py#L12

This way I think the `_file_exists_under_dir()` method will only be needed in the base file. Not sure how good this explanation was, so let me know if I need to clarify

Reply:
That makes sense thank you